src/daft-io/src/google_cloud.rs

```diff
@@ -192,6 +186,102 @@ impl GCSClientWrapper {
             }
         }
     }
+
+    async fn _ls_impl(
+        &self,
+        client: &Client,
+        bucket: &str,
+        key: &str,
+        delimiter: Option<&str>,
+        continuation_token: Option<&str>,
+    ) -> super::Result<LSResult> {
+        let req = ListObjectsRequest {
+            bucket: bucket.to_string(),
+            prefix: Some(key.to_string()),
+            end_offset: None,
+            start_offset: None,
+            page_token: continuation_token.map(|s| s.to_string()),
+            delimiter: Some(delimiter.unwrap_or("/").to_string()), // returns results in "directory mode"
+            max_results: Some(1000), // Recommended value from API docs
+            include_trailing_delimiter: Some(false), // This will not populate "directories" in the response's .item[]
+            projection: None,
+            versions: None,
+        };
+        let ls_response = client
+            .list_objects(&req)
+            .await
+            .context(UnableToListObjectsSnafu {
+                path: format!("gs://{}/{}", bucket, key),
+            })?;
+        let mut files = ls_response.items.map_or(vec![], |items| {
+            items
+                .into_iter()
+                .filter_map(|obj| {
+                    if obj.name.ends_with('/') {
+                        // Sometimes the GCS API returns "folders" in .items[], so we manually filter here
+                        None
+                    } else {
+                        Some(FileMetadata {
+                            filepath: format!("gs://{}/{}", bucket, obj.name),
+                            size: Some(obj.size as u64),
+                            filetype: FileType::File,
+                        })
+                    }
+                })
+                .collect::<Vec<_>>()
+        });
+        let mut dirs = ls_response.prefixes.map_or(vec![], |prefixes| {
+            prefixes
+                .into_iter()
```

Comment:
it would be better to extend this into the `files` vec directly

Reply:
Fixed the code to do iterator chaining into a single `.collect()` which is probably a little cleaner!